api/api.go

```diff
@@ -0,0 +1,127 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul-terraform-sync/event"
+)
+
+const (
+	defaultAPIVersion = "v1"
+
+	// StatusHealthy is the healthy status. This is determined based on status
+	// type.
+	//
+	// Task Status: Determined by the success of a task updating. The 5 most
+	// recent task updates are stored as an ‘event’ in CTS. A task is healthy
+	// when all the stored events are successful.
+	StatusHealthy = "healthy"
+
+	// StatusDegraded is the degraded status. This is determined based on status
+	// type.
+	//
+	// Task Status: Determined by the success of a task updating. The 5 most
+	// recent task updates are stored as an ‘event’ in CTS. A task is degraded
+	// when more than half of the stored events are successful _or_ less than
+	// half of the stored events are successful but the most recent event is
+	// successful.
+	StatusDegraded = "degraded"
+
+	// StatusCritical is the critical status. This is determined based on status
+	// type.
+	//
+	// Task Status: Determined by the success of a task updating. The 5 most
+	// recent task updates are stored as an ‘event’ in CTS. A task is critical
+	// when less than half of the stored events are successful and the most
+	// recent event is not successful
+	StatusCritical = "critical"
+
+	// StatusUndetermined is when the status is unknown. This is determined
+	// based on status type.
+	//
+	// Task Status: Determined by the success of a task updating. The 5 most
+	// recent task updates are stored as an ‘event’ in CTS. A task is
+	// undetermined when no event data has been collected yet.
+	StatusUndetermined = "undetermined"
+)
+
+// API supports api requests to the cts biniary
+type API struct {
+	store   *event.Store
+	port    int
+	version string
+	srv     *http.Server
+}
+
+// NewAPI create a new API object
+func NewAPI(store *event.Store, port int) *API {
+	mux := http.NewServeMux()
+
+	// retrieve task status for a task-name
+	mux.Handle(fmt.Sprintf("/%s/%s/", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(store, defaultAPIVersion))
+	// retrieve all task statuses
+	mux.Handle(fmt.Sprintf("/%s/%s", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(store, defaultAPIVersion))
+
+	srv := &http.Server{
+		Addr:         fmt.Sprintf(":%d", port),
+		WriteTimeout: time.Second * 15,
+		ReadTimeout:  time.Second * 15,
+		IdleTimeout:  time.Second * 60,
+		Handler:      mux,
+	}
+
+	return &API{
+		port:    port,
+		store:   store,
+		version: defaultAPIVersion,
+		srv:     srv,
+	}
+}
+
+// Serve starts up and handles shutdown for the http server to serve
+// API requests
+func (api *API) Serve(ctx context.Context) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				ctxShutDown, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

Comment:
I'd think you'd want to pass the context object `ctx` (the aurgment) instead of `context.Background()` to `context.WithTimout()` so that if the passed in context was canceled, the wrapped one would be as well. Generally this is the pattern, that as  you add new context at deeper levels that you always wrap the parent context. Used this way contexts keep the hierarchical aspect, so when you cancel one you know that all contexts under it also get canceled.

Reply:
I had tried passing in context i.e. `context.WithTimeout(ctx, 5*time.Second)` and noticed that when I stopped cts, cts would immediately shut down the server with a 'context canceled' and would not wait for open requests to finish. Looking in `Shutdown()` it returns as soon as it receives `<-ctx.Done()`. I think wrapping the parent context makes a lot of sense but isn't respecting the timeout as I thought it would. I might be missing something - let me know!

Reply:
I believe your findings are correct Lorna. The argument `ctx` is already closed and so the `Shutdown` functionality wouldn't proceed if taking that closed `ctx`.

Waiting on the `exitCh` 2x in `cli.go` accounts for this to wait on the server shutdown + timeout.

Reply:
Thanks for explanation @findkim. That makes sense! @eikenb let me know if you have any remaining thoughts on this

Reply:
Ok, this makes sense to me now as well. The usual pattern doesn't apply. :+1: